netroute_plan9.go

```diff
@@ -0,0 +1,141 @@
+// Generate a local routing table structure following the code at
+// https://github.com/google/gopacket/blob/master/routing/routing.go
+//
+// Plan 9 networking is described here: http://9p.io/magic/man2html/3/ip
+
+package netroute
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/google/gopacket/routing"
+)
+
+const netdir = "/net"
+
+func New() (routing.Router, error) {
+	rtr := &router{}
+	rtr.ifaces = make(map[int]net.Interface)
+	rtr.addrs = make(map[int]ipAddrs)
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, fmt.Errorf("could not get interfaces: %v", err)
+	}
+	for _, iface := range ifaces {
+		rtr.ifaces[iface.Index] = iface
+		var addrs ipAddrs
+		ifaceAddrs, err := iface.Addrs()
+		if err != nil {
+			return nil, err
+		}
+		for _, addr := range ifaceAddrs {
+			if inet, ok := addr.(*net.IPNet); ok {
+				// Go has a nasty habit of giving you IPv4s as ::ffff:1.2.3.4 instead of 1.2.3.4.
+				// We want to use mapped v4 addresses as v4 preferred addresses, never as v6
+				// preferred addresses.
+				if v4 := inet.IP.To4(); v4 != nil {
+					if addrs.v4 == nil {
+						addrs.v4 = v4
+					}
+				} else if addrs.v6 == nil {
+					addrs.v6 = inet.IP
+				}
+			}
+		}
+		rtr.addrs[iface.Index] = addrs
+	}
+
+	rtr.v4, rtr.v6, err = parseIPRoutes()
+	if err != nil {
+		return nil, err
+	}
+	return rtr, nil
+}
+
+func parseIPRoutes() (v4, v6 routeSlice, err error) {
+	buf, err := ioutil.ReadFile(netdir + "/iproute")
```

Comment:
i'm not familiar with p9 security. is it expected that most user processes will be able to read this successfully?
the system call structure of other implementations may provide a limited subset but tends not to fail for less-privileged processes. Not sure if that needs to be considered more here.

Reply:
We don't need to worry about it on Plan 9. If the process can make a TCP connection, it can read `/net/iproute`. Go's `net` package for Plan 9 is implemented entirely by file operations in `/net`, and the kernel device `#I` is normally mounted there.

The only reason it might fail is if the process is sanboxed to not have any networking, but that's a really special case.